parlai/mturk/tasks/acute_eval/run.py

```diff
@@ -0,0 +1,544 @@
+            conversation_task_list = conversations_to_tasks[id]
+            conversation_task_list.append(id)
+
+
+def get_new_task_data(worker, tasks_per_hit):
+    """ Get next task for worker. Returns the next onboarding task if worker
+    hasn't finished them all, or finds a task from the queue they haven't seen
+    If they've seen everything in the queue, spin up an extra task (one that
+    was in the queue and is now saturated)
+    """
+    worker_id = worker.worker_id
+    task_data = get_onboarding_tasks(worker_id, tasks_per_hit)
+    if len(task_data) == tasks_per_hit:
+        return onboarding_tasks
+    tries = 0
+    completed_tasks = workers_tasks_completed.get(worker_id, [])
+    seen_conversations = workers_to_conversations_seen.get(worker_id, [])
+    while (not task_queue.empty()) and tries < task_queue.qsize():
+        try:
+            next_task = task_queue.get()
+        except Queue.Empty:
+            break
+        tries += 1
+        if (  # make sure worker has not seen these conversations before
+            next_task['task_specs']['internal_id'] not in completed_tasks
+            and next_task['conversations'][0] not in seen_conversations
+            and next_task['conversations'][1] not in seen_conversations
+        ):
+            completed_tasks.append(next_task['task_specs']['internal_id'])
+            workers_tasks_completed[worker_id] = completed_tasks
+            seen_conversations.extend(next_task['conversations'])
+            workers_to_conversations_seen[worker_id] = seen_conversations
+            task_data.append(next_task)
+            if len(task_data) == tasks_per_hit:
+                return task_data
+        else:
+            task_queue.put(next_task)
+    # task queue is exhausted
+    tasks_still_needed = tasks_per_hit - len(task_data)
+    tasks_remaining = [id for id in desired_tasks.keys() if id not in completed_tasks]
+    tasks_chosen = [
+        t
+        for t in tasks_remaining
+        if desired_tasks[t]['conversations'][0] not in seen_conversations
+        and desired_tasks[t]['conversations'][1] not in seen_conversations
+    ]
+    if tasks_still_needed < len(tasks_chosen):
+        tasks_chosen = np.random.choice(tasks_chosen, tasks_still_needed, replace=False)
+    completed_tasks.extend(tasks_chosen)
+    seen_conversations.extend(
+        [desired_tasks[t]['conversations'][0] for t in tasks_chosen]
+    )
+    seen_conversations.extend(
+        [desired_tasks[t]['conversations'][1] for t in tasks_chosen]
+    )
+    task_data.extend([desired_tasks[id] for id in tasks_chosen])
+    return task_data
+
+
+def return_task_data(worker_id, task_data):
+    """ When worker doesn't complete a task, return it to the queue or
+    change their onboarding status depending on the task"""
+    for subtask_data in task_data:
+        if subtask_data['task_specs'].get('is_onboarding', False):
+            worker_onboarding_tasks_todo[worker_id].append(
```

Comment:
this is not defined. i think you meant `workers_to_onboarding_tasks_todo`. causes an error.